### ISSUE TITLE
linkerd should send absolute URI instead of full URI to the proxied service

### ISSUE THREAD

**Issue by L1AN0** [linkerd/linkerd2#5069]
### What is the issue?

Linkerd should send absolute URI instead of full URI to the proxied service

For example, if the proxied service runs

```
netcat -l -p 8000
```

and is after the service `example.example.svc.cluster.local`.

Then 

```
curl example.example.svc.cluster.local:8000
```

should have linkerd send the following request to the proxied service

```
GET / HTTP/1.1
Host: localhost:8000
User-Agent: curl/7.72.0
Accept: */*
```

instead of the current behavior (the following)

```
GET http://example.example.svc.cluster.local:8000/ HTTP/1.1
Host: localhost:8000
User-Agent: curl/7.72.0
Accept: */*
```

(See https://curl.haxx.se/mail/lib-2017-02/0107.html, and https://curl.haxx.se/mail/lib-2017-02/0108.html for why the first one is the desired behavior).

Not sending the absolute URI (`/`) but sending the full URI (`http://example.example.svc.cluster.local:8000/`) to the proxied service will result in failure in certain applications, including Python's default http server implementation and qbittorrent (e.g. https://hub.docker.com/r/linuxserver/qbittorrent). For example, if we inject linkerd to a python simple http server:

```
python3 -m http.server
```

accessing it will result in a 404 error (the following error on the server side):

```
127.0.0.1 - - [13/Oct/2020 02:10:37] "GET http://example.example.svc.cluster.local:8000/ HTTP/1.1" 404 -
```

while directly accessing the service works fine (the following log on the server side):

```
127.0.0.1 - - [13/Oct/2020 02:10:59] "GET / HTTP/1.1" 200 -
```

### How can it be reproduced?

See the description above.

### Logs, error output, etc

(If the output is long, please create a [gist](https://gist.github.com/) and
paste the link here.)

See the description above.

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ proxy-injector webhook has valid cert
√ sp-validator webhook has valid cert

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-addons
--------------
√ 'linkerd-config-addons' config map exists

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version:
- Cluster Environment: 
- Host OS: 
- Linkerd version: edge-20.10.2

**Comment by grampelberg** [comment]
I just tested this with `nc` and `curl` and am not getting the behavior you're seeing. Could you maybe provide more replication steps? I'm pretty surprised by this behavior in general as we don't modify that at all in the proxy.

With both the client and server meshed, I see:

```
curl frontend.test.svc.cluster.local:8081
```

```
# nc -l -p 8081
GET / HTTP/1.1
host: frontend.test.svc.cluster.local:8081
user-agent: curl/7.61.0
accept: */*
l5d-dst-canonical: frontend.test.svc.cluster.local:8081
```

**Comment by L1AN0** [comment]
I use the following yaml:

```
---
apiVersion: v1
kind: Namespace
metadata:
  name: issue-5069
---
apiVersion: v1
kind: Service
metadata:
  namespace: issue-5069
  name: netcat
spec:
  ports:
    - protocol: TCP
      name: web
      port: 8080
  selector:
    app: netcat
---
kind: Deployment
apiVersion: apps/v1
metadata:
  namespace: issue-5069
  name: netcat
  labels:
    app: netcat
spec:
  replicas: 1
  selector:
    matchLabels:
      app: netcat
  template:
    metadata:
      labels:
        app: netcat
      annotations:
        linkerd.io/inject: enabled
    spec:
      containers:
        - name: netcat
          image: subfuzion/netcat:latest
          imagePullPolicy: Always
          args: ["-lp", "8080"]
          ports:
            - name: web
              containerPort: 8080
---
kind: Deployment
apiVersion: apps/v1
metadata:
  namespace: issue-5069
  name: curl
  labels:
    app: curl
spec:
  replicas: 1
  selector:
    matchLabels:
      app: curl
  template:
    metadata:
      labels:
        app: curl
      annotations:
        linkerd.io/inject: enabled
    spec:
      containers:
        - name: curl
          image: ubuntu:latest
          imagePullPolicy: Always
          command: ["tail", "-f", "/dev/null"]
```

Then on the `curl` pod when I run `curl http://netcat.issue-5069.svc.cluster.local:8080` the server log shows:

```
GET http://netcat.issue-5069.svc.cluster.local:8080/ HTTP/1.1
host: netcat.issue-5069.svc.cluster.local:8080
user-agent: curl/7.68.0
accept: */*
l5d-dst-canonical: netcat.issue-5069.svc.cluster.local:8080
```

The cluster runs the latest k3s

```
k3s --version
k3s version v1.18.9+k3s1 (630bebf9)
```

@grampelberg

**Comment by grampelberg** [comment]
Looks like a regression in edge. It is working correctly in stable right now.

**Comment by olix0r** [comment]
@L1AN0 Thank you for the very helpful repro config! That made it much easier to identify and fix the underlying issue.

This fix may not land in today's edge release, but it will be released before stable-2.9.

**Comment by L1AN0** [comment]
Thank you for the prompt fix :)

**Comment by olix0r** [comment]
@L1AN0 The fix is included in `edge-20.10.3`, which is being released right now.

If you're impatient, you can try these changes now by adding workload annotations like:
```
config.linkerd.io/proxy-image: ghcr.io/olix0r/l2-proxy
config.linkerd.io/proxy-version: v2.115.0
```

I'm going to close this issue, but please don't hesitate to reopen it or to open a new issue if you find any problems! Thanks.
